python/tsfile/tsfile_pywrapper.pyx

```diff
@@ -0,0 +1,330 @@
+        cdef ErrorCode err_code
+        err_code = 0
+        self.writer = ts_writer_open(pathname.encode('utf-8'), &err_code)
+        if (err_code != 0):
+            raise Exception("Failed to open tsfile: %s, %s" %( pathname, err_code))
+    
+    def resister_timeseries(self, table_name, column_name, data_type):
+        cdef char* c_columns
+        cdef bytes py_table_name
+        cdef ColumnSchema schema
+        cdef bytes encoded_column_name = column_name.encode('utf-8')
+        py_table_name = table_name.encode('utf-8')
+        c_table_name = PyBytes_AsString(py_table_name)
+        schema.name = encoded_column_name
+        schema.column_def = data_type
+        if tsfile_register_table_column(self.writer, c_table_name, &schema) != 0:
+            raise Exception("Failed to register timeseries")
+    cdef create_row_data(self, table_name, time, column_length):
+        self.row_data = create_tsfile_row(table_name.encode('utf-8'), time, column_length)
+    def write_into_row_data(self, column_name, value, type):
+        cdef char* c_column_name = PyBytes_AsString(column_name.encode('utf-8'))
+        if type == TS_TYPE_INT32:
+            insert_data_into_tsfile_row_int32(self.row_data, c_column_name, value)
+        elif type == TS_TYPE_BOOLEAN:
+            insert_data_into_tsfile_row_boolean(self.row_data, c_column_name, value)
+        elif type == TS_TYPE_FLOAT:
+            insert_data_into_tsfile_row_float(self.row_data, c_column_name, value)
+        elif type == TS_TYPE_DOUBLE:
+            insert_data_into_tsfile_row_double(self.row_data, c_column_name, value)
+        elif type == TS_TYPE_INT64:
+            insert_data_into_tsfile_row_int64(self.row_data, c_column_name,  value)
+        else:
+            raise TypeError("Unknown column type")
+    def write_tsfile(self, table_name, df):
+        column_names = df.columns.tolist()
+        column_types = df.dtypes
+        column_ctypes = []
+        for i in range(len(column_names)):
+            column_type = column_types[i].name
+            if column_type in type_mapping:
+                column_ctypes.append(type_mapping[column_type])
+            else:
+                raise TypeError("Unknown column type")
+            
+            if (column_names[i] != TIMESTAMP_STR):
+                self.resister_timeseries(table_name, column_names[i], column_ctypes[i])
+
+                
+        for i in range(len(df)):
+            time = df.iloc[i][TIMESTAMP_STR]
+            self.create_row_data(table_name, time, len(column_names))
+            for j in range(1, len(column_names)):
+                column_name = column_names[j]
+                column_value = df.iloc[i][column_name]
+                column_ctype = column_ctypes[j]
+                self.write_into_row_data(column_name, column_value, column_ctype)
+            if tsfile_write_row_data(self.writer, self.row_data) != 0:
+                raise Exception("Failed to write row data")
```

Comment:
Writing a tablet in a row-based manner also has efficiency issues. We should add a todo for this.